### ISSUE TITLE
CORE-2 — Config system skeleton + model metadata schema

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1556]
**Depends on:** INF-1 (#1550) · **Blocks:** CFG-1 (#1574), CLI-1 (#1579), FM-3 (#1588)

**Context:** Today the entire training configuration surface is one flat argparse namespace: `mace/tools/arg_parser.py` is 1417 lines declaring **210 argparse dests across its two parsers, counted per parser** — 184 on `build_default_arg_parser` (`:14`) and 26 on `build_preprocess_arg_parser` (`:1239`), 188 of them distinct since 22 are declared in both — carried by 220 option strings, each count including the configargparse `--config` file flag (`:22-27`, `:1247-1252`, registered with `parser.add` rather than `add_argument`). A YAML file **is** a first-class input path — `configargparse` is a hard install dependency (`setup.cfg:32`), not an optional one — but there is no typed validation: bad keys and bad values surface deep inside the run, and what a model was actually trained with is not recorded in the model itself. The closest legacy gets is the TorchScript side artifact, which embeds `commit.txt` and a reflected `config.yaml` as `_extra_files` (`mace/cli/run_train.py:1184-1189`); the plain `.model` checkpoint that everyone actually loads is a bare pickle of the module (`:1183`) with no config, no provenance and no schema version. v1 rebuilds configuration around **Pydantic + pydantic-settings**: TOML/YAML/JSON input, **config-file-first** with dotted CLI overrides, unknown keys as hard errors, and the **fully resolved config (all defaults filled) saved into model metadata — mandatory for ALL models**, not just foundation models (decided; nearly free once this machinery and CLI-1 (#1579) exist). This ticket builds the *machinery* — the base config class and the metadata schema — not the full training schema: CFG-1 (#1574) later defines the real `data/model/loss/optimizer/schedule/runtime/finetune` sections on top of it, CLI-1 (#1579) wires the hierarchical `mace train/eval/model/data/export` CLI (including exporting the stored resolved config back to YAML and automatic citation printing), and FM-3 (#1588)'s model registry consumes `ModelMetadata` (DOIs, recommended/deprecated labels, query API). Skim `mace/tools/arg_parser.py` only for scale — do **not** port it; the legacy flag surface is mapped later (CFG-1 (#1574)/CLI-1 (#1579)).

**Interface & constraints:**

1. **`mace_core.config.ReforgeBaseConfig`** (pydantic-settings base class all v1 configs subclass):
   - Loads a config file by extension: `.toml`, `.yaml`/`.yml`, `.json` — same content in any of the three formats must produce an identical config object.
   - **Dotted CLI overrides** for arbitrary nested fields: `--model.num_interactions 3` overrides `model.num_interactions` from the file.
   - **Precedence, lowest to highest: defaults < config file < CLI override.** This is not a new decision — it is the precedence legacy already implements and pins (`tests/unit/test_arg_parser.py:135-145`: an explicit CLI flag beats the YAML value while untouched YAML values still apply). Tested explicitly here too, not assumed.
   - **Unknown keys are hard errors** that name the offending key *and its nearest valid neighbor* (agent-friendly errors: name the key/value and the fix; never silently ignore). This *is* a change from legacy, where configargparse accepts a YAML key for any declared flag and nothing validates the rest — declare it as such.
   - Hierarchical/nested structure is first-class — a demo subclass with two-level nesting is part of the deliverable, exercised in tests.
2. **Resolved export:** `to_resolved_dict()` returns the configuration with *all* defaults filled in, **stable key ordering**, and round-trippable: writing the resolved dict back to a file and loading it again reproduces itself. Property test: `file → resolved → file → resolved` is a **fixed point**. This resolved form is what gets embedded in model metadata and what a separate tool (CLI-1 (#1579)) later exports to YAML.
3. **`mace_core.metadata.ModelMetadata`** (Pydantic) — the mandatory per-model record:

| Field group | Content |
|---|---|
| Config | user-set config as explicitly given **and** the final resolved config with all defaults |
| Provenance | code version + git commit (legacy already collects the commit for the compiled artifact only — `run_train.py:1185`; v1 puts it in the checkpoint) |
| Data summary | automated summary of fitting data; documented convention that reference-quantity keys encode the calculation method (e.g. `pbe_energy`, `r2scan_forces`) |
| E0 details | explicit vs reestimated, plus the method/reference parameters used |
| Identity | DOI for the model |
| Citations | structured required citations |
| Free text | additional notes |

   - JSON round-trip with a **schema-version field**; an unknown *future* schema version is rejected with a clear message (never mis-parsed).
   - Citations render to a printable block via a **plain function** (no CLI machinery here — CLI-1 (#1579) consumes it for automatic citation printing).
4. **Framework-free:** `mace_core` imports neither torch nor jax; `mace_core.config` and `mace_core.metadata` must keep that invariant.

**Task:**
1. **`mace_core.config`:** implement `ReforgeBaseConfig` per the contract above — extension-based TOML/YAML/JSON loading, dotted CLI overrides, defaults < file < CLI precedence, unknown-key hard errors with nearest-neighbor suggestion.
2. **Resolved export:** `to_resolved_dict()` with all defaults filled, stable key ordering, round-trippable; property test that file→resolved→file→resolved is a fixed point.
3. **`mace_core.metadata`:** `ModelMetadata` with the field groups above; JSON round-trip + schema-version handling (reject unknown future versions with a clear message); citation-rendering function.
4. Demo config subclass exercising two-level nesting in tests.

**Out of scope:** the full training schema and cross-section validation (CFG-1 (#1574)); CLI entry points and the `mace` subcommand tree (CLI-1 (#1579)); the model registry / loaders that consume the metadata (FM-3 (#1588)); any port of the legacy argparse flag surface.

**Acceptance criteria:**
- [ ] The same config loads identically from TOML, YAML, and JSON (selected by file extension).
- [ ] Precedence is tested explicitly: a field set in defaults, file, and CLI ends up with the CLI value; defaults < file verified independently. The result matches the legacy precedence pinned at `tests/unit/test_arg_parser.py:135-145`.
- [ ] A dotted override (`--model.num_interactions 3`) reaches a two-level-nested field on the demo subclass.
- [ ] An unknown key raises a hard error naming the key and its nearest valid neighbor.
- [ ] `to_resolved_dict()` has all defaults filled and stable key ordering; the file→resolved→file→resolved fixed-point property test is green.
- [ ] `ModelMetadata` round-trips through JSON; the schema-version field is present; an unknown future schema version is rejected with a clear message.
- [ ] Structured citations render to a printable block via a plain function.
- [ ] `mace_core.config` and `mace_core.metadata` import neither torch nor jax.

**Inventory gaps assigned here:**
- [ ] Reproduce, as v1 tests over `ReforgeBaseConfig`, the two config-file behaviours legacy now pins (`tests/unit/test_arg_parser.py:117-145`): file values are applied, and an explicit CLI flag beats the file value while untouched file values still apply. The gap this row used to record — "`--config` works only through optional configargparse and no test covers it" — is closed on develop: configargparse is a required dependency and both behaviours are tested. The work here is inheriting that contract, not re-discovering it.
- [ ] Metadata in the **checkpoint**, not only in the side artifact: `ModelMetadata` must carry the commit + resolved config that legacy embeds solely in the `_compiled.model` `_extra_files` (`run_train.py:1184-1189`) — and which therefore vanishes whenever the silent `except Exception: pass` around that emission fires (`:1173-1174`, `:1188-1189`).

**Verify:**
```bash
python -m pytest packages/mace-core/tests/test_config.py packages/mace-core/tests/test_metadata.py -v
```
**Review focus:** the override/precedence semantics and the unknown-key error contract (CFG-1 (#1574) and CLI-1 (#1579) build directly on them — API churn here is expensive), and the metadata schema-versioning story: forward-compat rejection must be loud and the resolved-config embedding must be exactly the fixed-point form.

---
